ui/src/app/subnets/views/FormActions/components/AddSubnet.tsx

```diff
@@ -0,0 +1,149 @@
+import { Row, Col, Input } from "@canonical/react-components";
+import { useDispatch, useSelector } from "react-redux";
+import * as Yup from "yup";
+
+import type { FormActionProps } from "../FormActions";
+
+import FabricSelect from "app/base/components/FabricSelect";
+import FormikField from "app/base/components/FormikField";
+import FormikForm from "app/base/components/FormikForm";
+import VLANSelect from "app/base/components/VLANSelect";
+import { actions as subnetActions } from "app/store/subnet";
+import subnetSelectors from "app/store/subnet/selectors";
+import { toFormikNumber } from "app/utils";
+
+type AddSubnetValues = {
+  vlan: string;
+  name: string;
+  cidr: string;
+  gateway_ip: string;
+  dns_servers: string;
+  fabric: string;
+};
+
+const addSubnetSchema = Yup.object()
+  .shape({
+    cidr: Yup.string().required("CIDR is required"),
+    name: Yup.string(),
+    fabric: Yup.number(),
+    vlan: Yup.number(),
+    dns_servers: Yup.string(),
+    gateway_ip: Yup.string(),
+  })
+  .defined();
+
+const AddSubnet = ({
+  activeForm,
+  setActiveForm,
+}: FormActionProps): JSX.Element => {
+  const dispatch = useDispatch();
+  const isSaving = useSelector(subnetSelectors.saving);
+  const isSaved = useSelector(subnetSelectors.saved);
+  const errors = useSelector(subnetSelectors.errors);
+
+  return (
+    <FormikForm<AddSubnetValues>
+      validationSchema={addSubnetSchema}
+      buttonsBordered={false}
+      allowAllEmpty
+      initialValues={{
+        vlan: "",
+        name: "",
+        cidr: "",
+        gateway_ip: "",
+        dns_servers: "",
+        fabric: "",
+      }}
+      onSaveAnalytics={{
+        action: "Add Subnet",
+        category: "Subnets form actions",
+        label: "Add Subnet",
+      }}
+      submitLabel={`Add ${activeForm}`}
+      onSubmit={({ cidr, name, fabric, vlan, dns_servers, gateway_ip }) => {
+        dispatch(
+          subnetActions.create({
+            cidr,
+            name,
+            fabric: toFormikNumber(fabric),
+            vlan: toFormikNumber(vlan),
+            dns_servers,
+            gateway_ip,
+          })
+        );
+      }}
+      onCancel={() => setActiveForm(null)}
+      onSuccess={() => setActiveForm(null)}
+      saving={isSaving}
+      saved={isSaved}
+      errors={errors}
+    >
+      {({ values }) => (
```

Comment:
We generally try to avoid this way of working with Formik as you're effectively defining a component here, but without the benefits of a separate component (for example this will result in a new function being defined for every render of the parent component).

Where we need to access formik values or other formik functions we create a new `...Fields` component .e.g. https://github.com/canonical-web-and-design/maas-ui/blob/master/ui/src/app/machines/components/MachineHeaderForms/AddMachine/AddMachineForm/AddMachineForm.tsx#L200

Then we access the values using `useFormikContext`: https://github.com/canonical-web-and-design/maas-ui/blob/master/ui/src/app/machines/components/MachineHeaderForms/AddMachine/AddMachineFormFields/AddMachineFormFields.tsx#L27

Reply:
Thanks! Yeah, I used `useFormikContext` in a few places already. I can see the potential downsides of render props approach but also how it can be very convenient and simplify the implementation.

For now I will follow your suggestion for consistency.